include/deal.II/matrix_free/operators.h

```diff
@@ -834,6 +939,183 @@ namespace MatrixFreeOperators
+  void
+  LaplaceOperator<dim, fe_degree, n_q_points_1d, n_components, Number>::
+  set_coefficient(const std_cxx11::shared_ptr<Table<2, VectorizedArray<Number> > > &scalar_coefficient_ )
+  {
+    scalar_coefficient = scalar_coefficient_;
+  }
+
+
+
+  template <int dim, int fe_degree, int n_q_points_1d, int n_components, typename Number>
+  void
+  LaplaceOperator<dim, fe_degree, n_q_points_1d, n_components, Number>::
+  compute_diagonal()
+  {
+    Assert((Base<dim, Number>::data != NULL), ExcNotInitialized());
+
+    unsigned int dummy = 0;
+    LinearAlgebra::distributed::Vector<Number> &inverse_diagonal_entries = Base<dim,Number>::inverse_diagonal_entries;
+    this->initialize_dof_vector(inverse_diagonal_entries);
+    if (scalar_coefficient == NULL)
+      Base<dim,Number>::
+      data->cell_loop (&LaplaceOperator::local_diagonal_cell<true>,
+                       this, inverse_diagonal_entries, dummy);
+    else
+      Base<dim,Number>::
+      data->cell_loop (&LaplaceOperator::local_diagonal_cell<false>,
+                       this, inverse_diagonal_entries, dummy);
+
+    const std::vector<unsigned int> &
+    constrained_dofs = Base<dim,Number>::data->get_constrained_dofs();
+    for (unsigned int i=0; i<constrained_dofs.size(); ++i)
+      inverse_diagonal_entries.local_element(constrained_dofs[i]) = 0.;
+    for (unsigned int i=0; i<Base<dim,Number>::edge_constrained_indices.size(); ++i)
+      inverse_diagonal_entries.local_element(Base<dim,Number>::edge_constrained_indices[i]) = 0.;
+
+    for (unsigned int i=0; i<inverse_diagonal_entries.local_size(); ++i)
+      if (std::abs(inverse_diagonal_entries.local_element(i)) > std::sqrt(std::numeric_limits<Number>::epsilon()))
+        inverse_diagonal_entries.local_element(i) = 1./inverse_diagonal_entries.local_element(i);
+      else
+        inverse_diagonal_entries.local_element(i) = 1.;
+
+    Base<dim, Number>::inverse_diagonal_entries.compress(VectorOperation::insert);
+    Base<dim, Number>::inverse_diagonal_entries.update_ghost_values();
+  }
+
+
+
+  template <int dim, int fe_degree, int n_q_points_1d, int n_components, typename Number>
+  void
+  LaplaceOperator<dim, fe_degree, n_q_points_1d, n_components, Number>::
+  apply_add (LinearAlgebra::distributed::Vector<Number>       &dst,
+             const LinearAlgebra::distributed::Vector<Number> &src) const
+  {
+    if (scalar_coefficient == NULL)
+      Base<dim, Number>::data->cell_loop (&LaplaceOperator::local_apply_cell<true>,
+                                          this, dst, src);
+    else
+      Base<dim, Number>::data->cell_loop (&LaplaceOperator::local_apply_cell<false>,
+                                          this, dst, src);
+
+  }
+
+  namespace
+  {
+    template<typename Number>
+    bool
+    non_negative(const VectorizedArray<Number> &n)
+    {
+      bool res = true;
+      for (unsigned int v=0; v<VectorizedArray<Number>::n_array_elements; ++v)
+        res = res && (n[v] >= 0.);
```

Comment:
Do you think it is better with a fixed loop length? I would have done it like this which is slightly more obvious IMHO:
```
for (unsigned int v=0; v<VectorizedArray<Number>::n_array_elements; ++v)
￼  if (n[v] < 0)
    return false;
return true;
```
I wonder if a compiler would be able to optimize one better than the other. Anyway, since this is only used in debug code I do not think we need to care.